component/mimir/rules/rules.go

```diff
@@ -0,0 +1,309 @@
+	promClient   promVersioned.Interface
+	ruleLister   promListers.PrometheusRuleLister
+	ruleInformer cache.SharedIndexInformer
+
+	namespaceLister   coreListers.NamespaceLister
+	namespaceInformer cache.SharedIndexInformer
+	informerStopChan  chan struct{}
+	ticker            *time.Ticker
+
+	queue         workqueue.RateLimitingInterface
+	configUpdates chan ConfigUpdate
+
+	namespaceSelector labels.Selector
+	ruleSelector      labels.Selector
+
+	currentState map[string][]mimirClient.RuleGroup
+
+	metrics *metrics
+}
+
+type metrics struct {
+	configUpdatesTotal prometheus.Counter
+
+	eventsTotal   *prometheus.CounterVec
+	eventsFailed  *prometheus.CounterVec
+	eventsRetried *prometheus.CounterVec
+
+	mimirClientTiming *prometheus.HistogramVec
+}
+
+func (m *metrics) Register(r prometheus.Registerer) error {
+	r.MustRegister(
+		m.configUpdatesTotal,
+		m.eventsTotal,
+		m.eventsFailed,
+		m.eventsRetried,
+		m.mimirClientTiming,
+	)
+	return nil
+}
+
+func newMetrics() *metrics {
+	return &metrics{
+		configUpdatesTotal: prometheus.NewCounter(prometheus.CounterOpts{
+			Subsystem: "mimir_rules",
+			Name:      "config_updates_total",
+			Help:      "Total number of times the configuration has been updated.",
+		}),
+		eventsTotal: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Subsystem: "mimir_rules",
+			Name:      "events_total",
+			Help:      "Total number of events processed, partitioned by event type.",
+		}, []string{"type"}),
+		eventsFailed: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Subsystem: "mimir_rules",
+			Name:      "events_failed_total",
+			Help:      "Total number of events that failed to be processed, even after retries, partitioned by event type.",
+		}, []string{"type"}),
+		eventsRetried: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Subsystem: "mimir_rules",
+			Name:      "events_retried_total",
+			Help:      "Total number of retries across all events, partitioned by event type.",
+		}, []string{"type"}),
+		mimirClientTiming: prometheus.NewHistogramVec(prometheus.HistogramOpts{
+			Subsystem: "mimir_rules",
+			Name:      "mimir_client_request_duration_seconds",
+			Help:      "Duration of requests to the Mimir API.",
+			Buckets:   instrument.DefBuckets,
+		}, instrument.HistogramCollectorBuckets),
+	}
+}
+
+type ConfigUpdate struct {
+	args Arguments
+	err  chan error
+}
+
+var _ component.Component = (*Component)(nil)
+var _ component.DebugComponent = (*Component)(nil)
+
+func NewComponent(o component.Options, c Arguments) (*Component, error) {
+	setDefaultArguments(&c)
+
+	metrics := newMetrics()
+	metrics.Register(o.Registerer)
+
+	return &Component{
+		log:           o.Logger,
+		opts:          o,
+		args:          c,
+		configUpdates: make(chan ConfigUpdate),
+		ticker:        time.NewTicker(c.SyncInterval),
+		metrics:       metrics,
+	}, nil
+}
+
+func (c *Component) Run(ctx context.Context) error {
+	err := c.startup(ctx)
+	if err != nil {
+		return err
+	}
```

Comment:
I'm not sure about this returning an error; it'll terminate the component and won't spin back up. Should it log instead and set the health of the component to unhealthy so people can fix? 

Ditto with returning an error in the for loop.

Reply:
I think I incorrectly assumed returning an error would terminate the pod (meaning normal CrashLoop monitoring would alert someone). I'll make the change